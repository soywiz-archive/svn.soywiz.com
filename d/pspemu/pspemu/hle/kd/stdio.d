module pspemu.hle.kd.stdio; // kd/stdio.prx (sceStdio)

import pspemu.hle.Module;

class StdioForUser : Module {
	void initNids() {
		mixin(registerd!(0x172D316E, sceKernelStdin));
		mixin(registerd!(0xA6BAB2E9, sceKernelStdout));
		mixin(registerd!(0xF78BA90A, sceKernelStderr));
	}

	/**
	 * Function to get the current standard in file no
	 * 
	 * @return The stdin fileno
	 */
	SceUID sceKernelStdin() { return STDIN; }

	/**
	 * Function to get the current standard out file no
	 * 
	 * @return The stdout fileno
	 */
	SceUID sceKernelStdout() { return STDOUT; }

	/**
	 * Function to get the current standard err file no
	 * 
	 * @return The stderr fileno
	 */
	SceUID sceKernelStderr() { return STDERR; }
}

class StdioForKernel : StdioForUser {
}

enum : SceUID {
	STDIN  = -1,
	STDOUT = -2,
	STDERR = -3
}

static this() {
	mixin(Module.registerModule("StdioForKernel"));
	mixin(Module.registerModule("StdioForUser"));
}