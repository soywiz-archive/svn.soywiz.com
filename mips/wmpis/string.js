String.prototype.ltrim = function() {
	return this.replace(/^\s+/, '');
};

String.prototype.rtrim = function() {
	return this.replace(/\s+$/, '');
};

String.prototype.trim = function() {
	return this.ltrim().rtrim();
};

String.prototype.toInteger = function() {
	var str = this, radix = 10;
	if (str.substr(0, 1) == '\'') return str.charCodeAt(1);
	if (str.substr(0, 1) == '0') {		
		switch (this.substr(1, 1)) {			
			case 'x': radix = 16; str = this.substr(2); break; // Hexadecimal
			case 'b': radix =  2; str = this.substr(2); break; // Binary
			default:  radix =  8; str = this.substr(1); break; // Octal
		}
	}
	return parseInt(str, radix);
};

String.prototype.parseString = function() {
	eval('var v = ' + this + ';');
	return v;
};

Number.prototype.toHex = function(padding) {
	var h = '0123456789ABCDEF', r = '', n = (this >= 0) ? this : -this;
	while (n) { r = h.substr(n % 16, 1) + r; n = Math.floor(n / 16); }
	while (r.length < padding) r = '0' + r;		
	if (this < 0) r = '-' + r;
	return r;
};