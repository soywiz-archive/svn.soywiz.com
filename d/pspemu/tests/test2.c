#include <pspkernel.h>
#include <pspdisplay.h>
#include <pspdebug.h>
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
//#include <assert.h>

#include <pspctrl.h>
#include <pspgu.h>
#include <psprtc.h>

PSP_MODULE_INFO("Test2", 0, 1, 1);
PSP_MAIN_THREAD_ATTR(THREAD_ATTR_USER);

typedef struct {
	unsigned mantissa : 23;
	unsigned exponent : 8;
	unsigned sign     : 1;
} FLOAT;

#define assert(v) { if (!(v)) { asm("break"); } }
void emitInt     (int   v) { asm("syscall 0x2308"); }
void emitFloat   (float v) { asm("syscall 0x2309"); }
void emitString  (char *v) { asm("syscall 0x230A"); }
void startTracing()        { asm("syscall 0x230B"); }
void stopTracing()         { asm("syscall 0x230C"); }

void testIntToFloat() {
	int table[] = {8, 10, 12};
	int n;
	float x = 0;

	emitFloat(x);
	for (n = 0; n < sizeof(table) / sizeof(table[0]); n++) {
		x += table[n];
		emitFloat(x);
	}
}

void testFloatManip() {
	float f1 = -8.4;
	float f2 = +4.8;
	FLOAT ff;

	ff = *(FLOAT *)&f1;
	emitInt(ff.sign);
	emitInt(ff.exponent);
	emitInt(ff.mantissa);

	ff = *(FLOAT *)&f2;
	emitInt(ff.sign);
	emitInt(ff.exponent);
	emitInt(ff.mantissa);
}

void testFloatExtra() {
	float y = 0.;
	float x = 16.4;
	int n = 0;

	y = (float)frexp(x, &n );
	emitFloat(y);
	emitInt(n);
	//assert(y == 0.512500);
	//assert(n == 5);
}

void testBitOps() {
	unsigned int value = 0x73A4C5F0;
	unsigned int v1 = 1;
	unsigned int v2 = 1;
	v2 += 2;
	emitInt((value >> 7) & 0x7);
	emitInt(v1 ^ v2);
}

void testBitExtract() {
	unsigned int value = 0xC493A10F;
	int n;
	emitString("normal");
	for (n = 0; n < 32; n++) {
		emitInt((value >> n) & 1);
	}
	emitString("reversed");
	for (n = 31; n >= 0; n--) {
		emitInt(((value << n) & 0x80000000) >> 31);
	}
	assert(0);
}

void testDouble() {
	float f = 240.0f;
	double d;
	d = (double)f;
	emitFloat((float)d);
}

/*void testDtoa() {
	char data[16];
	int dsgn;
	int decpt = 2;
	char *rve;
	char *digits = _dtoa_r((void *)data, 240.0, 3, 6, &decpt, &dsgn, &rve);
	emitString("testDtoa");
	emitString(digits);
}*/

int main(int argc, char* argv[]) {
	//testDtoa();
	testDouble();
	testIntToFloat();
	testFloatManip();
	testFloatExtra();
	testBitOps();
	testBitExtract();

	return 0;
}