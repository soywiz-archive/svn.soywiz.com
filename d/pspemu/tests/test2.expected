240.000000
0.000000
8.000000
18.000000
30.000000
1
130
419430
0
129
1677722
0.512500
5
3
2
"normal"
1
1
1
1
0
0
0
0
1
0
0
0
0
1
0
1
1
1
0
0
1
0
0
1
0
0
1
0
0
0
1
1
"reversed"
1
1
1
1
0
0
0
0
1
0
0
0
0
1
0
1
1
1
0
0
1
0
0
1
0
0
1
0
0
0
1
1